### PR TITLE
add support for nacl SealedBox issue_41831

### PR DESCRIPTION
**Pull request by steverweber**
### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/41831

### New Behavior
add support for nacl SealedBox

### Tests written?
No
```
set -x

salt-call --local nacl.keygen sk_file=/tmp/nacl
data='neverhaveiever'
data_enc=`salt-call --local --out=newline_values_only nacl.enc "$data"`
salt-call nacl.dec "$data_enc"
unlink /tmp/nacl
unlink /tmp/nacl.pub

set +x
```

### CODE REVIEW

**Comment by steverweber** [comment]
@yagnik feel free to give it a go.

**Comment by steverweber** [comment]
This PR is a WIP... I'll rebase after feedback and some fixes

File: salt/modules/nacl.py [hunk 1]
```diff
@@ -29,80 +29,85 @@
...
 
 .. code-block:: jinja
 
     pillarexample:
         user: root
-        password: {{ salt.nacl.dec('DRB7Q6/X5gGSRCTpZyxS6hXO5LnlJIIJ4ivbmUlbWj0llUA+uaVyvou3vJ4=') }}
+        password1: {{salt.nacl.dec('DRB7Q6/X5gGSRCTpZyxS6hlbWj0llUA+uaVyvou3vJ4=')|json}}
+        password2: {{salt.nacl.dec_pub('CTpZyxS6hXO54ivbmUlbGSRlJIIJWj0llou3vJ4=')|json}}
 
-Or do something interesting with grains like:
```
Comment (cachedout) [inline salt/modules/nacl.py]: Why remove this example?

File: salt/modules/nacl.py [hunk 2]
```diff
@@ -160,6 +171,26 @@ def _get_key(rstrip_newline=True, **kwargs):
...
+def _get_key_pub(rstrip_newline=True, **kwargs):
+    '''
+    Return key pub
+    '''
+    config = _get_config(**kwargs)
+    key_pub = config['key_pub']
+    keyfile_pub = config['keyfile_pub']
+    if not key_pub and keyfile_pub:
+        if not os.path.isfile(keyfile_pub):
+            raise Exception('file not found: {0}'.format(keyfile_pub))
```
Comment (cachedout) [inline salt/modules/nacl.py]: Why not just catch an exception if the `fopen` fails to avoid the stat overhead every time?

**Comment by yagnik** [comment]
@steverweber thanks for working on this! I'll play around and introduce my renderer with this code. Will ping you on pr by tomorrow.

File: salt/modules/nacl.py [hunk 3]
```diff
@@ -184,17 +215,90 @@ def keygen(keyfile=None):
...
+    CLI Examples:
+
+    .. code-block:: bash
+
+        salt-call --local nacl.keygen_pub
+        salt-call --local nacl.keygen_pub key="YmFkcGFzcwo="
+        salt-call --local nacl.keygen_pub keyfile=/root/.nacl  keyfile_pub=/root/.nacl.pub
+        salt-call --local --out=newline_values_only nacl.keygen_pub > /root/.nacl.pub
+    '''
+    key = _get_key(**kwargs)
```
Comment (yagnik) [inline salt/modules/nacl.py]: In case key is not found, we should generate a key

**Review by yagnik — COMMENTED** [review]
minor comment

File: salt/modules/nacl.py (hunk 1 above)
Comment (steverweber) [inline salt/modules/nacl.py]: was too advanced and could open security issues.

**Comment by cachedout** [comment]
@steverweber Could you please fix these two minor lint issues? https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/12234/violations/

**Comment by steverweber** [comment]
@yagnik @cachedout 

Should SealedBox be made the new default...
i'm thinking of changing `nacl.dec_pub` to `nacl.dec` and the original `nacl.dec` to `nacl.secretbox_dec`
`nacl.dec will alias nacl.sealedbox_dec`
also change nacl.keygen to by default generate public and private keys.

**Comment by yagnik** [comment]
@steverweber I personally prefer explicit function names so not nacl.dec but nacl.sealedbox_dec and nacl.secretkey_dec instead f just nacl.dec. WDYT ?

+1 to one function to generate pub + priv keys

**Comment by steverweber** [comment]
> +1 to one function to generate pub + priv keys

nacl.keygen_pub does that,

anyways i'm just going to leave it as is.

this PR is ready.

**Comment by steverweber** [comment]
added support for
`{{salt.nacl.dec(instance.get('user', None))|default('xyz')}}`

https://github.com/saltstack/salt/issues/42003#issuecomment-312247638

**Comment by epcim** [comment]
I think we have to add dependency "libnacl" to requirements/opt.txt (etc..)

**Comment by steverweber** [comment]
recreated this PR... with a fresh branch because i had some messy merges with upstream and failing test.

**Pull request closed without merging**